### PR TITLE
Worker2

### PR DESCRIPTION
**Pull request by jonasfj**
can I get a 30% review on this..

I think I could replace the `atomics.Barrier` primitive with `context.Context` it just feels a bit abusive..
Note: I'm adding functions concurrency primitivies in the atomics package to try and avoid making the code too complicated elsewhere...

hmm.. the more I play with it in my mind... I'm starting to think I have to create a non-generic concurrency primitive that does what `activeTasks` and `idleTimer` tries to do...

### CODE REVIEW

**Comment by jonasfj** [comment]
@petemoore, please just look at the files in the `worker/` folder for 30% review...
The rest is also suggested in other PRs... it's sort of a mess, I tried to file in separate PRs, but also needed it all in one to write this... and didn't really want to land without review on the other PRs :) hehe
Chicken and their eggs..

File: worker2/worker.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		}
+
+		// If we received zero claims or encountered an error, we wait at-least
+		// pollingInterval before polling again. We start the timer here, so it's
+		// counting while we wait for capacity to be available.
+		var delay <-chan time.Time
+		if claims == nil || len(claims.Tasks) == 0 {
+			delay = time.After(w.pollingInterval)
+		} else {
+			delay = time.After(0)
```
Comment (petemoore) [inline worker2/worker.go]: I believe task claiming process could run in less time than `pollingInterval`, in which case, we could end up polling more frequently than `pollingInterval`. Also if multiple go routines were claiming work we'd be polling too frequently. Also this is code that would be easy to mess up, so any `0` delays are particularly dangerous as a bug could DDoS our services - would be better for `ClaimWork` to globally throttle, as a centralised safety net, and would also simplify code like this code above.

Comment (jonasfj) [inline worker2/worker.go]: We only sleep zero if we got tasks from the last `ClaimWork` call...

As long as we're getting new tasks, there is no need to sleep.

**Comment by jonasfj** [comment]
> I see why you created a Barrier primitive, I just wonder if we can do these things without introducing new primitives.

We definitely could do without introducing a new primitive... I could probably have used two `context.Context` objects... But I'm not sure it would be less confusing..

I make the new primitives because I hope I can hide logic in these primitives... So that the concurrency aspects are easier to understand when looking that code that does interesting things.
If barrier is hard to understand then I've definitely failed at that... So the question is did introduction of barrier make this simpler or more complicated?
@petemoore  ^ ?

**Comment by jonasfj** [comment]
Merging all my refactoring PRs into one big one... and integrating them...

It'll be horrible, but we can move forward..

**Pull request closed without merging**